src/Security/Member.php

```diff
@@ -1702,4 +1711,52 @@ public function getHtmlEditorConfigForCMS()
         // If can't find a suitable editor, just default to cms
         return $currentName ? $currentName : 'cms';
     }
+
+    /**
+     * This needs to be public because it's referenced in getMemberPasswordField() and called in ConfirmedPasswordField
+     */
+    public static function generateRandomPassword(array $args = []): string
+    {
+        $password = '';
+        $validator = self::password_validator();
+        $minLength = $validator ? $validator->getMinLength() : 20;
+        $maxLength = $args['maxLength'] ?? 0;
+        if ($maxLength === 0) {
+            $maxLength = max($minLength, 20);
+        } else {
+            if ($maxLength < $minLength) {
+                throw new InvalidArgumentException('maxLength argument is less than PasswordValidator minLength');
+            }
+        }
+        // Create a dummy member to test the password validator against
+        // Use a dummy member because this is intended to be used in a context where an Administrator
+        // is created a branch new member who doesn't yet have an ID
+        $dummyMember = Member::create();
+        // These are similar character sets that used in the default PasswordValidator
+        $charsets = [
+            'abcdefghijklmnopqrstuvwyxz',
+            'ABCDEFGHIJKLMNOPQRSTUVWYXZ',
+            '0123456789',
+            '!@#$%^&*()_+-=[]{};:,./<>?',
+        ];
+        $n = 0;
+        do {
+            $valid = true;
+            $password = '';
+            for ($i = 0; $i < $maxLength; $i++) {
+                $charset = $charsets[$i % 4];
+                $randomInt = random_int(0, strlen($charset) - 1);
+                $password .= $charset[$randomInt];
+            }
+            $password = str_shuffle($password);
+            if ($validator && !$validator->validate($password, $dummyMember)) {
+                $valid = false;
+            }
+            if (++$n === 100) {
+                // protection against infinite while loop
+                throw new RuntimeException('Unable to generate a random password');
+            }
+        } while (!$valid);
```

Comment:
Under what circumstances will this fail validation? The whole point of explicitly adding in each charset is to avoid having to have this loop and condition - we know what the validator validates against, so we can ensure our random password is always valid.

If people implement their own password validation logic, or random password will likely _always_ pass their custom validation, or _always_ fail it (we should probably add an extension hook or something here to let them update the logic)

Reply:
Have removed the while loop.

Reply:
Still calling validate though? My comment is still therefore valid and unanswered. When will this ever fail validation?

Reply:
Talked in person - this is so that if developers forget to implement the new extension hook they get a hard exception and can fix the problem before going to prod (ideally)